app/models/user.rb

```diff
@@ -2,9 +2,69 @@ class User < ApplicationRecord
   has_many :posts, dependent: :destroy
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships, foreign_key: "creator_id"
+  has_many :inverse_friendships, class_name: "Friendship", foreign_key: "recipient_id"
   belongs_to :location, required: false
+  validates :email, uniqueness: true
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: %i[facebook]
+
+
+  def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
+      end
+    end
+  end
+
+  def confirmed_friends
+    created_friendships = Array.new
+    created_friendships = friendships.select { |i| i.status == true}.map do |f|
+      f.recipient
+    end
+
+    received_friendships = Array.new
+    received_friendships = inverse_friendships.select { |i| i.status == true }.map do |f|
+    f.creator
+    end
+
+    (created_friendships + received_friendships)
```

Comment:
We have added logic to the model layer to prevent this from happening. In addition, we have checks on both the controller layer, and in the view layer to prevent this from happening.